fccodegen/_generator.py

```diff
@@ -158,39 +273,86 @@ def derivative_name(deriv):
         return 'd_dt_' + unames[var]
 
     # Create expression printer
-    p = cg.WebLabPrinter(symbol_name, derivative_name)
-
-    #TODO: Update cellmlmanip to make this code look more like:
-    #
-    # state_vars = model.get_state_variables()
-    # free_var = model.get_free_variable()
-    # derivatives = [sp.Derivative(state, free_var) for state in state_vars]
-    # eqs = model.get_equations_for(derivatives)
-    #
-
-    # Create a list of state variable indices, variables, and derivative vars
-    states = []
-    for i, state in enumerate(get_state_list(graph)):
-        states.append({
+    printer = cg.WebLabPrinter(symbol_name, derivative_name)
+
+    # Create free variable name
+    free_name = symbol_name(get_free_variable_symbol(graph))
+
+    # Create state information dicts
+    state_info = []
+    for i, state in enumerate(get_state_symbols(graph)):
+        #TODO: Update cellmlmanip so this graph.nodes goes away
+        initial_value = float(graph.nodes[state]['initial_value'])
+        state_info.append({
             'index': i,
             'var_name': symbol_name(state),
             'deriv_name': derivative_name(state),
+            'initial_value': initial_value,
+        })
+
+    # Create parameter information dicts
+    parameter_info = []
+    for i, parameter in enumerate(parameters):
+        symbol = get_symbol_by_cmeta_id(graph, parameter)
+        parameter_info.append({
+            'index': i,
+            'cmeta_id': parameter,
+            'var_name': symbol_name(symbol),
+            'initial_value': get_value(graph, symbol),
+        })
+
+    # Create map of parameter symbols to their indices
+    parameter_symbols = {}
+    for i, parameter in enumerate(parameters):
+        symbol = get_symbol_by_cmeta_id(graph, parameter)
+        parameter_symbols[symbol] = i
+
+    # Create output information dicts
+    # Each output is associated either with a symbol or a parameter.
+    output_info = []
+    for i, output in enumerate(outputs):
+        symbol = get_symbol_by_cmeta_id(graph, output)
+        output_info.append({
+            'index': i,
+            'cmeta_id': output,
+            'var_name': symbol_name(symbol),
+            'parameter_index': parameter_symbols.get(symbol, None),
         })
 
-    # Create the RHS equations
+    # Create RHS equation information dicts
     rhs_equations = []
-    for eq in eqs:
-        lhs, rhs = eq.lhs, eq.rhs
-        rhs_equations.append(p.doprint(lhs) + ' = ' + p.doprint(rhs))
+    for eq in get_equations_for(graph, get_derivative_symbols(graph)):
+        rhs_equations.append({
+            'lhs': printer.doprint(eq.lhs),
+            'rhs': printer.doprint(eq.rhs),
+            'parameter_index': parameter_symbols.get(eq.lhs, None),
```

Comment:
You shouldn't ever have (once the manipulation is done) a parameter occurring on the LHS of an equation. They can only be set from the protocol, not from a model equation.

Reply:
Might be my fault in cellmlmanip...?

Reply:
You have the right definition of a parameter for FC. However cellmlmanip will also be used to do the model manipulations specified by a protocol, so there will be some overlap. Once everything is connected up, the FC code would use cellmlmanip to modify the model equations, which will (inter alia) get rid of all equations with (FC) parameters on the LHS (they'll then only appear on the RHS of other equations), and then use fccodegen to generate the code to simulate the model.

Reply:
I'm still not sure how/where the manipulations will be applied! The protocol language seems to modify cellml, which would imply changes pre-parsing. Cellmlmanip removes cellml as soon as it can rather than actually manipulating it, right?

Reply:
The protocol language modifies equations, not CellML. So those changes happen post-parsing. (This is a key difference from SED-ML, which modifies the raw XML.)

Reply:
Thanks! Have made a note